File: go/lesson-09-testing/src/calculator.go
package calculator

import "fmt"

func Add(a, b float64) float64 {
	return a + b
}

func Subtract(a, b float64) float64 {
	return a - b
}

func Multiply(a, b float64) float64 {
	return a * b
}

func Divide(a, b float64) (float64, error) {
	if b == 0 {
		return 0, fmt.Errorf("cannot divide by zero")
	}
	return a / b, nil
}

// Factorial returns n! for non-negative n.
func Factorial(n int) (int, error) {
	if n < 0 {
		return 0, fmt.Errorf("factorial of negative number %d is undefined", n)
	}
	result := 1
	for i := 2; i <= n; i++ {
		result *= i
	}
	return result, nil
}

File: nodejs/lesson-09-testing/src/calculator.js
'use strict';

function add(a, b) { return a + b; }
function subtract(a, b) { return a - b; }
function multiply(a, b) { return a * b; }

function divide(a, b) {
  if (b === 0) throw new Error('Division by zero');
  return a / b;
}

function factorial(n) {
  if (!Number.isInteger(n) || n < 0) {
    throw new RangeError(`factorial requires a non-negative integer, got ${n}`);
  }
  if (n <= 1) return 1;
  return n * factorial(n - 1);
}

module.exports = { add, subtract, multiply, divide, factorial };
